internal/internal_workflow.go

```diff
@@ -1158,21 +1163,31 @@ func (s *coroutineState) close() {
 }
 
 // exit tries to run Goexit on the coroutine and wait for it to exit
-// within timeout.
-func (s *coroutineState) exit(timeout time.Duration) {
+// within timeout. If it doesn't exit within timeout, it will log a warning.
+func (s *coroutineState) exit(logger log.Logger, warnTimeout time.Duration) {
 	if !s.closed.Load() {
 		s.unblock <- func(status string, stackDepth int) bool {
 			runtime.Goexit()
 			return true
 		}
 
-		timer := time.NewTimer(timeout)
+		timer := time.NewTimer(warnTimeout)
 		defer timer.Stop()
 
 		select {
 		case <-s.aboutToBlock:
+			return
 		case <-timer.C:
+			st, err := getCoroStackTrace(s, "running", 0)
+			if err != nil {
+				st = fmt.Sprintf("<%s>", err)
+			}
+
+			logger.Warn(fmt.Sprintf("Workflow coroutine %q didn't exit within %v", s.name, warnTimeout), "stackTrace", st)
 		}
+		// We need to make sure the coroutine is closed, otherwise we risk concurrent coroutines running
+		// at the same time causing a race condition.
+		<-s.aboutToBlock
```

Comment:
How does this accomplish the comment above? If we receive a signal from `aboutToBlock` we exit cleanly, and otherwise if the timer fire first, why do we send a signal to `aboutToBlock` when the timer expires? 

Is it just to trigger the next coroutine exiting that it can exit cleanly?

Reply:
We don't send a signal to `aboutToBlock`, we wait on receiving a signal from `aboutToBlock`, which is sent on exit

Reply:
oops, misread this statement as sending a signal.. thanks for clarifying. So this function's behavior is the same, with an added log statement on timeout.

I'm a little confused by the PR title, this seems more like adding logs to help identify if evicted co-routines are timing out/exiting in the same order, but there is no code to actually make sure this is happening?

Reply:
It behaves a bit differently because previously we didn't have this second `<-s.aboutToBlock` to if the coroutine didn't exit fast enough we would end up evicting multiple coroutines in parallel. Now we always wait for `<-s.aboutToBlock` to fire before we continue

Reply:
You're right.. Maybe I need more coffee.. thanks for clarifying!